src/System.Data.Extensions/src/System/Data/DataTableExtensions.cs

```diff
@@ -0,0 +1,229 @@
+        /// with copies of the source rows.
+        /// Also note that this will cause the rest of the query to execute at this point in time
+        /// (e.g. there is no more delayed execution after this sequence operator).
+        /// </summary>
+        /// <param name="source">The input sequence of DataRows. CopyToDataTable uses DataRowVersion.Default when retrieving values from source DataRow
+        /// which will include proposed values for DataRow being edited. Null DataRow in the sequence are skipped.</param>
+        /// <param name="table">The target DataTable to load.</param>
+        /// <param name="options">The target DataTable to load.</param>
+        /// <param name="errorHandler">Error handler for recoverable errors.
+        /// Recoverable errors include:
+        ///   A source DataRow is in the deleted or detached state state.
+        ///   DataTable.LoadDataRow threw an exception, i.e. wrong # of columns in source row
+        /// Unrecoverable errors include:
+        ///   exceptions from IEnumerator, DataTable.BeginLoadData or DataTable.EndLoadData</param>
+        /// <returns>DataTable containing copies of the source DataRows.</returns>
+        /// <exception cref="ArgumentNullException">if source is null</exception>
+        /// <exception cref="ArgumentNullException">if table is null</exception>
+        /// <exception cref="InvalidOperationException">if source DataRow is in Deleted or Detached state</exception>
+        public static void CopyToDataTable<T>(this IEnumerable<T> source, DataTable table, LoadOption options, FillErrorEventHandler errorHandler)
+            where T : DataRow
+        {
+            DataSetUtil.CheckArgumentNull(source, "source");
+            DataSetUtil.CheckArgumentNull(table, "table");
+            LoadTableFromEnumerable(source, table, options, errorHandler);
+        }
+
+        private static DataTable LoadTableFromEnumerable<T>(IEnumerable<T> source, DataTable table, LoadOption? options, FillErrorEventHandler errorHandler)
+            where T : DataRow
+        {
+            if (options.HasValue)
+            {
+                switch (options.Value)
+                {
+                    case LoadOption.OverwriteChanges:
+                    case LoadOption.PreserveChanges:
+                    case LoadOption.Upsert:
+                        break;
+                    default:
+                        throw DataSetUtil.InvalidLoadOption(options.Value);
+                }
+            }
+
+
+            using (IEnumerator<T> rows = source.GetEnumerator())
+            {
+                // need to get first row to create table
+                if (!rows.MoveNext())
+                {
+                    if (table == null)
+                    {
+                        throw DataSetUtil.InvalidOperation(SR.DataSetLinq_EmptyDataRowSource);
+                    }
+                    else
+                    {
+                        return table;
```

Comment:
Nit: another example of something that could be cleaned up later to take advantage of newer C# features, e.g.
```C#
if (!rows.MoveNext())
{
    return table ?? throw DataSetUtil.InvalidOperation(SR.DataSetLinq_EmptyDataRowSource);
}
```